File: src/main/java/net/swade/permissionss/group/Group.java
package net.swade.permissionss.group;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

@Getter
@AllArgsConstructor
public class Group {
    private final String name, id, nameTagFormat, chatFormat;
    private final List<String> aliases;
    private final List<String> permissions;
    private final List<String> inheritance;

    public List<String> getAllPermissions() {
        List<String> perms = new ArrayList<>(this.permissions);
        if (inheritance.isEmpty()){
            return perms;
        }
        for (String inheritance : this.inheritance) {
            Group group = GroupManager.getGroup(inheritance);
            assert group != null;
            perms.addAll(group.getAllPermissions());
        }
        return perms;
    }
}

File: src/main/kotlin/group/Group.kt
package net.guneyilmaz0.permissions.group

data class Group(
    val name: String,
    val id: String,
    val nameTagFormat: String,
    val chatFormat: String,
    val aliases: List<String>,
    val permissions: List<String>,
    val inheritance: List<String>
) {
    fun getAllPermissions(): List<String> {
        val perms = permissions.toMutableList()
        inheritance.forEach { inherit ->
            GroupManager.getGroup(inherit)?.getAllPermissions()?.let { perms.addAll(it) }
        }
        return perms
    }
}